PFW-1212: Add Dutch language

3d-gussner opened the pull request:
This is an attempt to add additional languages to Prusa Firmware, starting with Dutch.

Why: As we have in our maker space in Belgium children and few people not that familiar with English and there are other Dutch speaking people might appreciate this, I started to translate the Prusa Firmware in June 2017 for the Prusa MK2. After some feedback from users we added French and Portuguese as well. The Pull Requests in the past didn't make it BUT i hope Prusa might consider to implement more languages. 

I also wrote basic `How-to add a new language to Prusa Firmware` so others may be able to add other languages.

PFW-1212

3d-gussner committed:
Prepare adding new language

3d-gussner committed:
Update documentation

3d-gussner committed:
Fix syntax

3d-gussner committed:
Looks nicer

3d-gussner committed:
Another syntax fix

3d-gussner committed:
Fix typo

3d-gussner committed:
more typo

3d-gussner committed:
Create new language files

3d-gussner committed:
Create new language files #2

3d-gussner committed:
Translated to Dutch

3d-gussner committed:
Update Dutch po files

3d-gussner committed:
Merge branch 'MK3' into MK3_Dutch

Merge with MK3 and fix merge issues

3d-gussner committed:
Fix translation file and odd chars in `lang-import.sh`

3d-gussner committed:
Update `lang_en_nl.txt` to latest PRs made in MK3 branch

3d-gussner committed:
Update `.po` files after modifying `lang_en*.txt` files

3d-gussner committed:
Merge branch 'MK3' into MK3_Dutch

Fix merge conflicts

3d-gussner committed:
Merge branch 'MK3' into MK3_Dutch

3d-gussner committed:
Updated Dutch translation to latest merge
some messages have been deleted

3d-gussner committed:
Merge branch 'MK3' into MK3_Dutch

3d-gussner committed:
Update Dutch translation to v3.8.1-RC1

3d-gussner committed:
Updated po files

3d-gussner committed:
Some more missing translation

3d-gussner committed:
Merge branch 'MK3' into MK3_Dutch

3d-gussner committed:
Updated `lang_en_nl.txt` with newest translations

3d-gussner committed:
Merge branch 'MK3' into MK3_Dutch

vintagepc commented on lang/po/Firmware_nl.po:
```diff
@@ -0,0 +1,1787 @@
+# 
+#: 
+msgid "[%.7s]Live adj. Z\x0avalue set, continue\x0aor start from zero?\x0a%cContinue%cReset"
+msgstr "[%.7s]Z instell.\x0awaarde ingesteld,\x0averder of bij 0\x0abeginen?\x0a%cVerder%cReset"
```
"terugvallen op 0" might be better than "bij 0 beginnen" for the intent of "revert to 0"

vintagepc commented on lang/po/Firmware_nl.po:
```diff
@@ -0,0 +1,1787 @@
+# MSG_AUTOLOADING_ONLY_IF_FSENS_ON c=20 r=4
+#: ultralcd.cpp:4459
+msgid "Autoloading filament available only when filament sensor is turned on..."
+msgstr "Autoloading filament alleen beschikbaar wanneer filamentsensor is ingeschakeld ..."
```
"Auto-laden" for consistency with the previous?

3d-gussner replied:
"Auto. laden van fil. is enkel beschikbaar wanneer fil.sensor is ingeschakeld ..."
This fits perfectly the 20x4

vintagepc commented on lang/po/Firmware_nl.po:
```diff
@@ -0,0 +1,1787 @@
+# MSG_BED_CORRECTION_MENU
+#: ultralcd.cpp:5838
+msgid "Bed level correct"
+msgstr "Bed niveau correct"
```
"vlak" of "waterpas" is better for "flatness" in the context of bed levelling; "niveau" is generally more used in the context of height.

vintagepc commented on lang/po/Firmware_nl.po:
```diff
@@ -0,0 +1,1787 @@
+# MSG_EJECTING_FILAMENT c=20 r=1
+#: mmu.cpp:1413
+msgid "Ejecting filament"
+msgstr "Werp filament uit"
```
Translation seems closer to "remove filament" than ejecting, if there's space I'd suggest
"Fil. aan het uitwerpen" ("filament is ejecting")

3d-gussner replied:
Is te lang met c=22 r=1

vintagepc replied:
"Fil. word ontladen"?

vintagepc commented on lang/po/Firmware_nl.po:
```diff
@@ -0,0 +1,1787 @@
+# MSG_WIZARD_WILL_PREHEAT c=20 r=4
+#: ultralcd.cpp:5080
+msgid "Now I will preheat nozzle for PLA."
+msgstr "Opwarmen van de tuit voor PLA voor."
```
"Voorverwarming" is probably better for preheat, e..g. "Voorverwarming van de tuit vor PLA"

vintagepc commented on lang/po/Firmware_nl.po:
```diff
@@ -0,0 +1,1787 @@
+# MSG_PRINT_PAUSED c=20 r=1
+#: ultralcd.cpp:1074
+msgid "Print paused"
+msgstr "Print pauzeren"
```
pauzeren is active, as in "pause print". Past tense ("paused") is "gepauzeerd"

vintagepc commented on lang/po/Firmware_nl.po:
```diff
@@ -0,0 +1,1787 @@
+# MSG_FIND_BED_OFFSET_AND_SKEW_LINE1 c=60
+#: messages.c:38
+msgid "Searching bed calibration point"
+msgstr "Zoeken bed kalibratiepunt"
```
would suggest "Bed kalibratiepunt aan het zoeken" if there is sufficient space.

jurriaan replied:
Bed isn't an adjective here, so it can be Bedkalibratiepunt I guess

vintagepc commented on lang/po/Firmware_nl.po:
```diff
@@ -0,0 +1,1787 @@
+# 
+#: ultralcd.cpp:2312
+msgid "to unload filament"
+msgstr "om filament te laden"
```
typo, should be "om filament uit te werpen"

3d-gussner replied:
Have to double check but i think this message is limited to c=20 r=1
Your translation is correct but is too long.

vintagepc replied:
"om fil. te ontladen" should fit, probably better

vintagepc commented on lang/po/Firmware_nl.po:
```diff
@@ -0,0 +1,1787 @@
+# MSG_XYZ_DETAILS c=19 r=1
+#: ultralcd.cpp:2227
+msgid "XYZ cal. details"
+msgstr "XZY kal. details"
```
XZY?

3d-gussner replied:
Typical typo... glad you reviewed it

vintagepc commented on lang/po/Firmware_nl.po:
```diff
@@ -0,0 +1,1787 @@
+# 
+#: ultralcd.cpp:7152
+msgid "Magnets comp.[N/A]"
+msgstr "Magnet. comp.[N/V]"
```
could use "Magneet" directly or "Magneten" for plural

3d-gussner replied:
Limited space

vintagepc reviewed (commented):
Reviewed and made some translation suggestions.

3d-gussner commented:
@vintagepc Thanks for the comments, I will review them.
It would be better to comment on the `lang_en_nl.txt` file as these files are used during firmware build. The `.po` files are ment to exchange translations between `code` and translators.
Another thing we have to keep in mind is that the LCD can only display 20 char in 4 rows.
One of my next tasks will be to add the missing c=xx r=yy comments to be able to see what limits we have.

This means:
- one line messages may not be longer than 20 chars
- menu messages are even more limited in length
- some status messages are really short
- it is sometimes hard to use correct plural
- some words are just too long (German has some really long words)
- having a multiple row message, we still need to see how it will be shown.
  - A long message will be split into rows while it tries to get as many as possible words per row, this can look very odd if you use very long words after each other.

vintagepc commented:
Understood, I knew we had space limitations but wasn't aware of the exact allowances for each case. I'm also quite aware that many EU languages can be very regional in choices of phrasing and syntax, further complicating the mess. :)

leptun commented:
There are also languages where if you keep shortening words, the message stops making sense. This is the case for Romanian. Things are also complicated by lack of proper translation counterpart. Most words either sound forced (the english counterpart is a better word that most people will understand and is the actual terminology used by the community) or are too long (ie 20+ chars for a full word).

vintagepc commented:
Arabic is a fun one to deal with in translation... there if you shorten a word, it can actually take up more characters than before!

leptun commented:
Luckily we don’t have to deal with these languages. The displays can’t display something like that unless it’s higher resolution and the cells are next to eachother (or there aren’t cells at all). Also, that is only an edge case, words usually get smaller in arabic just like in the other languages.

3d-gussner committed:
Merge remote-tracking branch 'upstream/MK3' into MK3_Dutch

DRracer commented on Firmware/language.c:
```diff
@@ -209,6 +209,7 @@ const char* lang_get_name_by_code(uint16_t code)
 	case LANG_CODE_ES: return _n("Espanol");
 	case LANG_CODE_FR: return _n("Francais");
 	case LANG_CODE_IT: return _n("Italiano");
+	case LANG_CODE_NL: return _n("Nederlands");
```
Please tag Dutch as community contribution ...

3d-gussner committed:
Merge remote-tracking branch 'upstream/MK3' into MK3_Dutch

Merge upstream/MK3
Fix conflicts
update Dutch translation

3d-gussner committed:
Updated few translations
thanks to @vintagepc and @jurriaan for the feedback

3d-gussner committed:
Tag as community contribution

3d-gussner commented:
@vintagepc @jurriaan Thanks for the feedback, I updated few translations.
Can you please review the `lang_en_nl.txt` and comment if something is wrong?

3d-gussner committed:
Merge remote-tracking branch 'upstream/MK3' into MK3_Dutch

fix merge issues

3d-gussner committed:
Translate new messages
Check that amount of messages are equal `lang_en.txt` vs. `lang_en_nl.txt`

3d-gussner committed:
Use new `PF-build.sh` parameter `-c 1`

3d-gussner committed:
Update workflow

3d-gussner commented:
@vintagepc Can you please review the changes https://github.com/prusa3d/Prusa-Firmware/pull/2223/commits/afa84b15bdaf09b837d091e60ab6613e943fb21d ?

3d-gussner commented:
> Maybe "De printer drukt nu een zigzaglijn. Draai..." ? Feels like it reads a little more naturally.

We have to choose between "print" and "drukt" to keep wording similar. There are over places where we use "print" instead of "drukt"

3d-gussner commented:
> Not sure if this is a regional dialect thing but I think "Riem" is more of a clothing-type belt; I've always heard "snaar" used for "technical" type belts like those in tape decks etc.

https://www.lexilogos.com/english/dutch_translation.htm 3x "riem" once "gordel"

DRracer commented:
@3d-gussner @vintagepc another almost finished feature ready to be tested. 
I'd really like to see this this up and running since we managed to overcome the 64K limit in the XFLASH and verified, that it works

3d-gussner committed:
Add defines and templates for community languages
Add `Community made` submenu to `Select language`

3d-gussner committed:
Add `lang-community.sh` script
Fix `Dutch` hex file copied in `PF-build.sh` for MK2.5/S

3d-gussner committed:
Move `Dutch` language parts
Not create community languages with `lang-build.sh all` the `lang-community.sh` will call it.
Add templates for future community languages

3d-gussner committed:
Update lang translation files

3d-gussner committed:
Update

3d-gussner committed:
Fix indentation

3d-gussner committed:
Add template

3d-gussner committed:
Update po files

3d-gussner committed:
Merge remote-tracking branch 'upstream/MK3' into MK3_Dutch

3d-gussner committed:
Fix back from submenu `Community made` to `Select language` after factory reset

3d-gussner committed:
Run `lang-community.sh` in build.sh
